### ISSUE TITLE
C++ gRPC client hangs occasionally when running against google bigtable

### ISSUE THREAD

**Issue by ay60** [grpc/grpc#6278]
I am running into an issue with my C++ gRPC client that hangs occasionally when running against google bigtable. The call that hangs most often is `ReadRows` which is a streaming call. I have seen the `MutateRow` call hanging occasionally as we but it is very rare.  The hanging is random yet somewhat reproducible, in a sense that the same single threaded application doing the same calls in the same order eventually hangs on one of the `ReadRows` call. With GRPC_TRACE flags on I can see these events:

```
I0424 15:25:33.611778339  539240 call.c:1473]                grpc_call_start_batch(call=0x7f40441288a0, ops=0x7f4051a56650, nops=2, tag=0x7f4051a56940, reserved=(nil))
I0424 15:25:33.611786813  539240 call.c:1185]                ops[0]: RECV_INITIAL_METADATA ptr=0x7f4051a56968
I0424 15:25:33.611808718  539240 call.c:1185]                ops[1]: RECV_MESSAGE ptr=0x7f4051a56990
I0424 15:25:33.611844407  539240 client_channel.c:109]       OP[client-channel:0x7f4044128e08]: RECV_INITIAL_METADATA RECV_MESSAGE
I0424 15:25:33.611854950  539240 http_client_filter.c:143]   OP[http-client:0x7f4044126be0]: RECV_INITIAL_METADATA RECV_MESSAGE
I0424 15:25:33.611861653  539240 connected_channel.c:70]     OP[connected:0x7f4044126bf8]: RECV_INITIAL_METADATA RECV_MESSAGE
D0424 15:25:33.611895742  539240 tcp_posix.c:401]            WRITE 0x7f4044121c60: 17 03 03 00 25 00 00 00 00 00 00 0a 82 67 54 d8 03 5d ab 3b b9 24 d0 3d 90 be 47 05 01 cb 6c 7e 3f cc bd fa 90 5f 2a 8f 44 f6 '....%........gT..].;.$.=..G...l~?...._*.D.'
I0424 15:25:33.611969989  539240 completion_queue.c:397]     grpc_completion_queue_pluck(cc=0x7f40441208a0, tag=0x7f4051a56940, deadline=gpr_timespec { tv_sec: 9223372036854775807, tv_nsec: 0, clock_type: 1 }, reserved=(nil))
D0424 15:25:33.615595350  539240 tcp_posix.c:163]            read: success=1
D0424 15:25:33.615608468  539240 tcp_posix.c:167]            READ 0x7f4044121c60: 17 03 03 00 27 00 00 00 00 00 00 6a e3 18 d0 b2 61 0f a3 30 8e a9 88 80 c4 bc a6 a9 5e f3 92 84 44 c7 23 11 06 88 db f1 29 8f 93 0b '....'......j....a..0........^...D.#.....)...'
I0424 15:25:43.614348296  539240 client_channel.c:109]       OP[client-channel:0x7f4044128e08]: CANCEL:4
I0424 15:25:43.614368378  539240 http_client_filter.c:143]   OP[http-client:0x7f4044126be0]: CANCEL:4
I0424 15:25:43.614373615  539240 connected_channel.c:70]     OP[connected:0x7f4044126bf8]: CANCEL:4
```

The cancel call at 15:25:43 is triggered by the deadline timeout. I am submitting this issue to gRPC team to help me to understand what is happening here. Clearly there is a response coming from the server but it seems to be insufficient for the gRPC code to take any actions. The `grpc_completion_queue_pluck` call fails with a timeout. The 10 seconds timeout value is somewhat random but in my experience that call never returns anyway, regardless of how long do I wait. The session is practically dead after the timeout kicks in. I had to make use of a new connection once that timeout happens.

Any suggestions? Is that a problem on the server side or on the client side? I can provide more information or do more targeted debugging if asked.

**Comment by ay60** [comment]
Here is a stack trace:

```
#0  0x00007f2b9c66712d in poll () at ../sysdeps/unix/syscall-template.S:81
#1  0x00007f2b9bcfb674 in basic_pollset_maybe_work_and_unlock (exec_ctx=0x7f2b9b22ea50, pollset=0x7f2b8c11fec8, worker=0x7f2b9b22e980, deadline=..., now=...) at src/core/iomgr/pollset_posix.c:579
#2  0x00007f2b9bcfad74 in grpc_pollset_work (exec_ctx=0x7f2b9b22ea50, pollset=0x7f2b8c11fec8, worker_hdl=0x7f2b9b22ea20, now=..., deadline=...) at src/core/iomgr/pollset_posix.c:307
#3  0x00007f2b9bd0bcbe in grpc_completion_queue_pluck (cc=0x7f2b8c11fde0, tag=0x7f2b9b22eb10, deadline=..., reserved=0x0) at src/core/surface/completion_queue.c:459
#4  0x00007f2b9c132f3a in grpc::CompletionQueue::Pluck (this=0x7f2b8c118798, tag=0x7f2b9b22eb10) at src/cpp/common/completion_queue.cc:80
#5  0x00007f2b9d42ec78 in grpc::ClientReader<google::bigtable::v1::ReadRowsResponse>::Read (this=0x7f2b8c118780, msg=0x7f2b8c12dc10) at /usr/local/include/grpc++/impl/codegen/sync_stream.h:150
```

There are similar stack traces provided with some issues. Typically the response from the gRPC team is that this is a server side issue, such as server threads starvation.  Since the server in this cases is bigtable and there is a response of some kind from the server seen in the logs, I am not sure if we can blame the server in this case.

**Comment by ay60** [comment]
Any updates?

**Comment by ctiller** [comment]
Which version of gRPC are you using?

Did you use GRPC_TRACE=all? I suspect not - I'm not seeing some bits I'd
expect to see - and it'd be useful for tracking this down.

**Comment by ay60** [comment]
@ctiller I was having this problem with all versions of gRPC up to this commit below. I have not tried it since. I make builds off the main branch, gRPC version is something like 0.14.

```
commit ab1c84f49506f86cb9590708e777d65a4a2a2f10
Merge: 9ea6027 896ae71
Author: Jan Tattermusch <jtattermusch@users.noreply.github.com>
Date:   Fri Apr 8 22:42:57 2016 -0700
```

The GRPC_TRACE options that I used are those:
```
GRPC_TRACE=api,channel,http,flowctl,connectivity_state,tcp
```
For some reasons `flowctl` did not show up in the last traces. Not sure why. It seems like it should.

I can try again with the `GRPC_TRACE=all` and the tip of the mainline. Do you think it's worth it to try again? Were there any fixes made in the relevant areas of code in the last month or two?

**Comment by yang-g** [comment]
I think some traces are only logged for debug builds and maybe that is why
the flowctl traces do not show up.

**Comment by ctiller** [comment]
It'd be worthwhile to do another trace: with just 'tcp' we're only seeing
encrypted bytes. 'all' or adding 'secure_endpoint' will let us see the
decrypted bytes also, and let us ascertain what HTTP2 bytes were put onto
the wire. 'all' also includes 'channel' which will let us see a little more
of what's going on in the gRPC stack.

**Comment by ay60** [comment]
Thanks! I'll give it another try today with `all` option and using the latest code too. Will do the debug build.

**Comment by ay60** [comment]
Here is the latest case. `GRPC_TRACE=all` was used. gRPC C++  code at
```
commit 0d6196025e62aea5aabc6341459f2c370e264230
Merge: 704fb3f 05e8c9a
Author: Craig Tiller <ctiller@google.com>
Date:   Tue May 31 10:20:20 2016 -0700
```

The call is still the same `ReadRows` Bigtable streaming call.
Note the 10 seconds time gap from 19:17:56 to 19:18:06. That's the hanging. The read call returns after 10 seconds deadline timeout. Let me know if you want to see more of the log.

```
I0531 19:17:56.279246413  113246 call.c:1504]                grpc_call_start_batch(call=0x7fe0e42588f0, ops=0x7fe0f2b2a650, nops=2, tag=0x7fe0f2b2a940, reserved=(nil))
I0531 19:17:56.279262421  113246 call.c:1215]                ops[0]: RECV_INITIAL_METADATA ptr=0x7fe0f2b2a968
I0531 19:17:56.279273483  113246 call.c:1215]                ops[1]: RECV_MESSAGE ptr=0x7fe0f2b2a990
I0531 19:17:56.279287567  113246 client_channel.c:109]       OP[client-channel:0x7fe0e4258e50]: RECV_INITIAL_METADATA RECV_MESSAGE
I0531 19:17:56.279318226  113246 http_client_filter.c:143]   OP[http-client:0x7fe0e4260ec0]: RECV_INITIAL_METADATA RECV_MESSAGE
I0531 19:17:56.279333258  113246 connected_channel.c:70]     OP[connected:0x7fe0e4260ed8]: RECV_INITIAL_METADATA RECV_MESSAGE
D0531 19:17:56.279349498  113246 chttp2_transport.c:1774]    FLOW       op: client stream[3657] CREDIT             global.max_recv_bytes(65535) by                add_max_recv_bytes(65535) giving 131070
D0531 19:17:56.279364301  113246 chttp2_transport.c:1777]    FLOW       op: client stream[3657] CREDIT global.unannounced_incoming_window_for_parse(0) by                add_max_recv_bytes(65535) giving 65535
D0531 19:17:56.279925298  113246 chttp2_transport.c:1780]    FLOW       op: client stream[3657] CREDIT global.unannounced_incoming_window_for_writing(0) by                add_max_recv_bytes(65535) giving 65535
D0531 19:17:56.279952770  113246 writing.c:146]              FLOW    write: client stream[3657] MOVE                 writing.announce_window(0) <- global.unannounced_incoming_window_for_writing(65535) giving 65535
D0531 19:17:56.279972259  113246 writing.c:225]              FLOW    write: client stream[3657] DEBIT            writing.announce_window(65535) by                          announce(65535) giving 0
D0531 19:17:56.279986834  113246 secure_endpoint.c:259]      WRITE 0x7fe0e4266690: 00 00 04 08 00 00 00 0e 49 00 00 ff ff '........I....'
D0531 19:17:56.280017150  113246 tcp_posix.c:401]            WRITE 0x7fe0e4259b20 (peer=ipv4:64.233.182.219:443): 17 03 03 00 25 00 00 00 00 00 00 11 32 84 34 37 89 da 8c 5a f7 e9 ae 1d ca 70 fb b5 47 62 77 be 14 75 c0 84 ff 2b 93 ab 65 2e '....%.......2.47...Z.....p..Gbw..u...+..e.'
I0531 19:17:56.280096101  113246 completion_queue.c:405]     grpc_completion_queue_pluck(cc=0x7fe0e41208b0, tag=0x7fe0f2b2a940, deadline=gpr_timespec { tv_sec: 9223372036854775807, tv_nsec: 0, clock_type: 1 }, reserved=(nil))
D0531 19:17:56.289603747  113246 tcp_posix.c:163]            read: success=1
D0531 19:17:56.289663844  113246 tcp_posix.c:167]            READ 0x7fe0e4259b20 (peer=ipv4:64.233.182.219:443): 17 03 03 00 27 00 00 00 00 00 01 43 ae f0 0b 74 af e2 95 e0 99 85 e0 23 22 41 f2 35 6f f6 92 ec 43 5f fc b4 8a 2b 78 f1 89 f7 11 8a '....'......C...t.......#"A.5o...C_...+x.....'
D0531 19:17:56.289712529  113246 secure_endpoint.c:135]      READ 0x7fe0e4266690: 00 00 06 01 04 00 00 0e 49 88 ea e9 be e7 e6 '........I......'
D0531 19:17:56.289755660  113246 parsing.c:98]               FLOW    parse: client stream[3657] MOVE             parsing.incoming_window(65535) <- global.unannounced_incoming_window_for_parse(65535) giving 131070
I0531 19:17:56.289776704  113246 parsing.c:617]              HTTP:3657:HDR:CLI: :status: 200
I0531 19:17:56.289784898  113246 parsing.c:617]              HTTP:3657:HDR:CLI: content-disposition: attachment
I0531 19:17:56.289790517  113246 parsing.c:617]              HTTP:3657:HDR:CLI: content-type: application/grpc
I0531 19:17:56.289796521  113246 parsing.c:617]              HTTP:3657:HDR:CLI: date: Wed, 01 Jun 2016 02:17:56 GMT
I0531 19:17:56.289824038  113246 parsing.c:617]              HTTP:3657:HDR:CLI: alternate-protocol: 443:quic
I0531 19:17:56.289831763  113246 parsing.c:617]              HTTP:3657:HDR:CLI: alt-svc: quic=":443"; ma=2592000; v="34,33,32,31,30,29,28,27,26,25"
I0531 19:18:06.277157856  113246 client_channel.c:109]       OP[client-channel:0x7fe0e4258e50]: CANCEL:4
I0531 19:18:06.277194349  113246 http_client_filter.c:143]   OP[http-client:0x7fe0e4260ec0]: CANCEL:4
I0531 19:18:06.277204201  113246 connected_channel.c:70]     OP[connected:0x7fe0e4260ed8]: CANCEL:4
```

**Comment by ay60** [comment]
@ctiller how is it going? That new trace, is it helpful at all? Any insights to what might be happening? I would be happy to try new things, like do more debugging or make a build off a branch to verify any fixes or attempts at if these are forthcoming. Just let me know.

**Comment by ctiller** [comment]
Hey... I missed this earlier... thanks for pinging.

The last read in the logs is here:
D0531 19:17:56.289712529  113246 secure_endpoint.c:135]      READ 0x7fe0e4266690: 00 00 06 01 04 00 00 0e 49 88 ea e9 be e7 e6 '........I......'

And shows a header frame of length 6 being returned and parsed correctly.

After that there's radio silence: it looks as though the server isn't returning anything.

**Comment by ay60** [comment]
Does that mean that this is Bigtable server issue? Do we need to get someone from the Bigtable team involved?

**Comment by ay60** [comment]
@ctiller I have a relevant question. In C++ code base, with the streaming calls, last time I looked, there was no way to set a timeout on an individual read call, only on the whole lifetime of the stream. There is no specific timeout on something like `ClientReader::Read()`, which is what the generated code is using. It is only the `ClientReader()` constructor that sets the deadline. I was going to file an issue asking for a change. I am mentioning that now because to deal with this specific Bigtable problem I am setting deadline but I can not tune that deadline value very well because I have to be conscious of the total possible lifetime of the `ClientReader`. Should I file a different issue? There really should be a way to set an individual timeout on every next `ClientReader::Read()` call.

**Comment by sduskis** [comment]
@ay60  and the cloud bigtable team had a discussion today.  A couple of interesting tidbits came up

1. Once the problem comes up, the "session" (in ay60's terms... I'm assuming that means a Channel) does not work for additional requests.  He has to create a new session in order to perform new RPCs.

2. @ay60 is creating a ReadRowRequest with a range starting at a particular key with an open ended stop key.  He reads some of the streamed responses, and then cancels the call.  We've had some problems in the past with similar requests where the response has not been fully consumed on the client side.  We've run tests with without canceling the call, and we've seen that many of these types of long running, partially consumed response RPCs tickles a server-side flow control issue which OOMs our servers. ay60's case could cause the same issue, and we'll have to test for it.

3. @ay60 uses creates a new channel once the old channel gets into a bad state, he gets a timeout after a few seconds and then reconnects and retries the query, which works.

We'll definitely do more testing about the "start a large scan, read a bit, cancel, repeat" scenario.  It could be another case that isn't handled effectively.

In the short term, perhaps the best way forward is to work on an approach that doesn't invoke the potential OOM.  That may involve a bit of Bigtable schema design which will allow better targeting of range scans.

**Comment by ay60** [comment]
Just to clarify. I am not sure if any of the scans that the app is doing can be qualified as "large scan". The test suite inserts the total of about 700 records in the database. The scan keys are not open-ended on the lower boundary. Still, even if they are, there are no more that 700 rows to read.

**Comment by sduskis** [comment]
I removed the word large.  The issue of "start a scan, read a few rows, cancel, repeat" still needs some investigation on our end.

**Comment by dibyendumajumdar** [comment]
Hi, I am facing a problem where my client hangs at CompletionQueueSafeHandle.Pluck(). I was looking for any reports of a similar problem and I found this, so it may or may not be related. I have described the problem I am facing in this post:

https://groups.google.com/forum/#!topic/grpc-io/1t5UC8lPnxs

I was trying to understand what might be causing the issue - and I wondered whether it is something to do with using pointers as tags. I don't understand what these tags are used for but is it possible that every so often the tags conflict? I may be talking nonsense here.

Regards
Dibyendu
